File: 152.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        f, g= [nums[0]], [nums[0]]
        for i in range(1, len(nums)):
            f.append(max(f[i-1]*nums[i], g[i-1]*nums[i], nums[i]))
            g.append(min(f[i-1]*nums[i], g[i-1]*nums[i], nums[i]))
        return max(f)

File: 152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        vector<int> f(n, 0), g(n,0);
        f[0] = nums[0];
        g[0] = nums[0];
        int res = nums[0];
        for (int i = 1; i < nums.size();i++){
            f[i] = max({nums[i], f[i-1] * nums[i], g[i-1] * nums[i]});
            g[i] = min({nums[i], f[i-1] * nums[i], g[i-1] * nums[i]});
            res = max(f[i], res);
        }
        return res;
    }
};
